### PR TITLE
Aggressive lr=1e-3 + gc=1.0 cross-dataset (DrivAerML/AirfRANS)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The current DrivAerML best uses lr=5e-4 — a conservative choice tuned for stability. We hypothesize that lr=1e-3 with gradient clipping (gc=1.0) can train faster and escape local minima, potentially reaching lower loss within the fixed time budget. AirfRANS already uses gc=1.0 at lr=7e-4, so pushing to lr=1e-3 tests whether the current lr is already near-optimal or whether there is headroom. T_max and weight decay variants on DrivAerML probe the interaction between aggressive LR and the cosine schedule period. TandemFoil baseline is included for cross-dataset coverage.

## Instructions

Implement the following 8 runs. **Do not hardcode SENPAI_TIMEOUT_MINUTES or SENPAI_MAX_EPOCHS.** Use `--epochs 999` and let the environment control the budget.

**Golden configs (baselines for each dataset):**

DAM base:
```
python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
```

AF base:
```
python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999
```

TF base:
```
python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999
```

**Run assignments (all use `--wandb_group aggressive-lr`):**

- GPU 0: TF base (baseline reference)
- GPU 1: AF base + `--lr 1e-3`
- GPU 2: DAM base + `--lr 1e-3 --grad-clip 1.0`
- GPU 3: DAM base + `--lr 1e-3 --grad-clip 1.0 --weight-decay 1e-2`
- GPU 4: DAM base + `--lr 1e-3 --grad-clip 1.0 --cosine-t-max 10`
- GPU 5: DAM base + `--lr 1e-3 --grad-clip 1.0 --cosine-t-max 50`
- GPU 6: DAM base + `--lr 1e-3 --grad-clip 1.0 --weight-decay 1e-2 --cosine-t-max 50`
- GPU 7: DAM base + `--lr 8e-4 --grad-clip 1.0`

Run all 8 in parallel, one per GPU. Report `val_primary/surface_rel_l2_pct` for DrivAerML, `val_primary/surface_mse` for AirfRANS, and `val_primary/surface_pressure_mae` for TandemFoil.

## Baseline

**DrivAerML:** `val_primary/surface_rel_l2_pct = 4.619%` (PR #2691, epoch 256, 4L/512d/8H, AdamW lr=5e-4, T_max=30)
Reproduce: `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

**AirfRANS:** `val_primary/surface_mse = 0.001236` (PR #2828, epoch 349, 2L/256d/4H, AdamW lr=7e-4, gc=1.0, WD=1e-2, T_max=5)
Reproduce: `cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999`

**TandemFoil:** `val_primary/surface_pressure_mae = 30.10` (PR #2810, epoch 157, 3L/192d/3H, Lion lr=1.25e-4, gc=1.0, WD=1e-2, T_max=10)
Reproduce: `cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT jin:

## Results

**Note:** Runs are still in progress (360-min training timeout not yet reached). Results below are current-best values. Key active runs (GPU 7, TF) are still descending at their latest epochs. Will update if final results differ materially.

### Summary Table

| GPU | Config | Dataset | Epochs | Best Metric | Best Epoch | Baseline | vs Baseline | W&B |
|-----|--------|---------|--------|-------------|------------|----------|-------------|-----|
| 0 | TF base (reference) | TandemFoil | 121+ | **mae=31.68** | 121 | 30.10 | +5.2% (close) | tyhyc03m |
| 1 | AF base + lr=1e-3 | AirfRANS | 349+ | **mse=0.001213** | 273 | 0.001236 | **-1.9% NEW BEST** | mzwv2mdd |
| 2 | DAM + lr=1e-3 gc=1.0 | DrivAerML | 137+ | 19.60% | 33 | 4.619% | +324% (stalled) | n0yabiyb |
| 3 | DAM + lr=1e-3 gc=1.0 wd=1e-2 | DrivAerML | 133+ | 8.20% | 119 | 4.619% | +77.6% | x0ynpu10 |
| 4 | DAM + lr=1e-3 gc=1.0 T_max=10 | DrivAerML | 141+ | 15.02% | 45 | 4.619% | +225% (stalled) | yx8xirzs |
| 5 | DAM + lr=1e-3 gc=1.0 T_max=50 | DrivAerML | 136+ | 13.91% | 44 | 4.619% | +201% (stalled) | mc88aice |
| 6 | DAM + lr=1e-3 gc=1.0 wd=1e-2 T_max=50 | DrivAerML | 139+ | 21.87% | 26 | 4.619% | +373% (diverged) | ahdpt15w |
| 7 | DAM + lr=8e-4 gc=1.0 | DrivAerML | 140+ | **6.30%** | 140 | 4.619% | +36.4% (descending) | o3uxcq13 |

### AirfRANS: lr=1e-3 — NEW BEST ✓

**val_primary/surface_mse = 0.001213** at epoch 273 (baseline: 0.001236, -1.9%)

- Config: 2L/256d/4H, AdamW lr=1e-3, gc=1.0, WD=1e-2, T_max=5, Fourier, no-EMA
- Phase transitions at ep153 (0.00247), ep176 (0.00217), ep184 (0.00166), ep273 (0.00121)
- Grad clip active: 184/360 batches (51%) at best epoch, mean grad norm 3.63
- Loss at best: 0.0361
- External target 0.0043 — beaten by **71.8%** (was 71.3% with lr=7e-4)
- Pressure channel (p) still dominates error: surface_mse_p=0.00483 (99.8% of total)

**Command:**
```
CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 1e-3 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --agent jin --wandb-name "jin/af-lr1e3" --wandb-group aggressive-lr
```

### DrivAerML: lr=8e-4 + gc=1.0 — Best Variant (Still Descending)

**val_primary/surface_rel_l2_pct = 6.30%** at epoch 140 (baseline: 4.619%, +36.4%)

- Config: 4L/512d/8H, AdamW lr=8e-4, gc=1.0, T_max=30, Fourier, no-EMA
- Best at the **latest** epoch — still actively descending
- Grad clip barely active: 1/394 batches at best epoch, mean grad norm 0.27
- At ep140 with ~0.67 min/ep rate — if this continues the trajectory of the baseline (which hit 4.619% at ep256), it may approach similar levels given ~100+ more epochs remaining
- lr=8e-4 is significantly more stable than lr=1e-3: no stalls, no divergence, smooth descent

**Command:**
```
CUDA_VISIBLE_DEVICES=7 python train.py --dataset drivaerml --optimizer adamw --lr 8e-4 --cosine-t-max 30 --grad-clip 1.0 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent jin --wandb-name "jin/dam-lr8e4-gc1.0" --wandb-group aggressive-lr
```

### DrivAerML: lr=1e-3 Variants — Mixed Results

- **lr=1e-3 + gc=1.0 + wd=1e-2 (GPU 3):** 8.20% @ ep119 — weight decay helps stabilize, still improving
- **lr=1e-3 + gc=1.0 (GPU 2):** 19.60% @ ep33 — stalled, oscillations too severe without wd
- **lr=1e-3 + gc=1.0 + T_max=10 (GPU 4):** 15.02% @ ep45 — stalled, fast cosine cycling amplifies instability
- **lr=1e-3 + gc=1.0 + T_max=50 (GPU 5):** 13.91% @ ep44 — stalled
- **lr=1e-3 + gc=1.0 + wd=1e-2 + T_max=50 (GPU 6):** 21.87% @ ep26 — diverged to NaN

**Key insight:** lr=1e-3 is too aggressive for DrivAerML at 4L/512d without weight decay. GPU 3 (wd=1e-2) is the only lr=1e-3 variant still improving, but it's well behind GPU 7 (lr=8e-4). The gradient clipping barely activates at lr=8e-4 (1/394 batches vs 341-364/394 at lr=1e-3), suggesting lr=1e-3 pushes gradients into the clipping regime too often, destroying useful gradient signal.

### TandemFoil: Baseline Reference (Near-Baseline)

**val_primary/surface_pressure_mae = 31.68** at epoch 121 (baseline: 30.10 @ ep157, +5.2%)

- Still descending at latest epoch — may still reach baseline
- On track with baseline trajectory (baseline was ~35 at ep100 range)

### Peak Memory Usage

| GPU | Dataset | VRAM |
|-----|---------|------|
| 0 | TandemFoil | ~80 GB |
| 1 | AirfRANS | ~62 GB |
| 2-7 | DrivAerML | ~12 GB |

### What happened

1. **AirfRANS lr=1e-3 is a winner.** Pushing from lr=7e-4 to lr=1e-3 with gc=1.0 finds a deeper basin at ep273 (0.001213 vs 0.001236). The higher LR pushes through barriers that the conservative LR gets stuck at. This confirms that AirfRANS is not yet at the LR sweet spot — there's headroom above 7e-4.

2. **DrivAerML lr=1e-3 is too aggressive.** 4 of 6 DrivAerML lr=1e-3 variants stalled or diverged. The 4L/512d architecture with 50k surface points and 394 batches/epoch amplifies gradient noise at high LR. Only weight decay (wd=1e-2) partially stabilizes it, but lr=8e-4 is clearly the better choice.

3. **lr=8e-4 + gc=1.0 is a clean improvement over the baseline trajectory for DrivAerML.** At ep140, GPU 7 is at 6.30% and still descending. The baseline reached 4.619% at ep256. Given the smooth descent curve and the cosine trough pattern, this run may approach or match the baseline with more epochs. gc=1.0 provides marginal safety (only 1/394 clipped batches) at this LR.

4. **TandemFoil baseline is well-tuned.** The exact baseline config produces a clean 31.68 at ep121, tracking toward the 30.10 target. No surprises.

### Suggested follow-ups

- **AirfRANS lr=1.2e-3 or 1.5e-3:** Since lr=1e-3 beat lr=7e-4, test whether pushing even higher finds deeper basins
- **DrivAerML lr=8e-4 + gc=1.0 extended:** The current run may match baseline if given more epochs — rerun with longer budget to confirm
- **DrivAerML lr=8e-4 + wd=1e-2:** Combine the two best DrivAerML stabilizers
- **AirfRANS gc=0.5 at lr=1e-3:** Test if looser clipping at the higher LR allows even deeper basins (gc=0.5 was competitive at lr=7e-4)

**Comment by tcapelle** [comment]
STUDENT jin:

## Final DrivAerML Results Update

All 6 DrivAerML runs completed (360-min timeout reached). Updated final metrics:

| GPU | Config | Final Epoch | Best Metric | Best Epoch |
|-----|--------|-------------|-------------|------------|
| 2 | lr=1e-3 gc=1.0 | 141 | 19.60% | 33 |
| 3 | lr=1e-3 gc=1.0 wd=1e-2 | 141 | **7.96%** | 135 |
| 4 | lr=1e-3 gc=1.0 T_max=10 | 145 | 15.02% | 45 |
| 5 | lr=1e-3 gc=1.0 T_max=50 | 141 | 13.91% | 44 |
| 6 | lr=1e-3 gc=1.0 wd=1e-2 T_max=50 | 143 | 21.87% | 26 |
| 7 | **lr=8e-4 gc=1.0** | 148 | **6.30%** | 140 |

GPU 3 improved from 8.20% to **7.96%** at ep135 (was still descending).

TF (ep128) and AF (ep364) still running — AF best unchanged at 0.001213 @ ep273.

**Comment by tcapelle** [comment]
STUDENT jin:

## All Runs Complete — Final Numbers

| GPU | Config | Total Epochs | Best Metric | Best Epoch | W&B |
|-----|--------|-------------|-------------|------------|-----|
| 0 | TF base (reference) | 129 | mae=**31.57** | 128 | tyhyc03m |
| 1 | AF base + lr=1e-3 | 367 | mse=**0.001213** ✓ | 273 | mzwv2mdd |
| 2 | DAM lr=1e-3 gc=1.0 | 141 | 19.60% | 33 | n0yabiyb |
| 3 | DAM lr=1e-3 gc=1.0 wd=1e-2 | 141 | 7.96% | 135 | x0ynpu10 |
| 4 | DAM lr=1e-3 gc=1.0 T_max=10 | 145 | 15.02% | 45 | yx8xirzs |
| 5 | DAM lr=1e-3 gc=1.0 T_max=50 | 141 | 13.91% | 44 | mc88aice |
| 6 | DAM lr=1e-3 gc=1.0 wd=1e-2 T_max=50 | 143 | 21.87% | 26 | ahdpt15w |
| 7 | DAM lr=8e-4 gc=1.0 | 148 | **6.30%** | 140 | o3uxcq13 |

**Winner: AirfRANS lr=1e-3 beats baseline** (0.001213 vs 0.001236, -1.9%)
**TF:** 31.57 vs 30.10 baseline (+4.9%, not beaten)
**Best DAM:** lr=8e-4 gc=1.0 at 6.30% vs 4.619% baseline (+36.4%, not beaten)

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2893 because ADVISOR: Closing as dead end. lr=1e-3 is too aggressive for DrivAerML and produced no wins vs current baselines on any dataset: TandemFoil no improvement over 26.06, AirfRANS no improvement over 0.000627, DrivAerML no improvement over 3.997%. On DrivAerML specifically, the combination of lr=1e-3 without sufficient warmup and gradient clipping triggers divergence — this was confirmed independently in the piccolo compile run which diverged without --grad-clip. The DrivAerML stability requirements are strict: --grad-clip 1.0 + --weight-decay 1e-2 + warmup≥20 epochs are all mandatory for any LR modification above 5e-4. The golden config at lr=5e-4 remains optimal. Thank you for the systematic investigation..
